python/ql/src/semmle/python/Files.qll

```diff
@@ -72,6 +72,33 @@ class File extends Container {
    * are specified to be extracted.
    */
   string getContents() { file_contents(this, result) }
+
+  /** Holds if this file is likely to get executed directly, and thus act as an entry point for execution. */
+  predicate maybeExecutedDirectly() {
+    // Only consider files in the source code, and not things like the standard library
+    exists(this.getRelativePath()) and
+    (
+      // The file doesn't have the extension `.py` but still contains Python statements
+      not this.getExtension() = "py" and
+      exists(Stmt s | s.getLocation().getFile() = this)
+      or
```

Comment:
Have we seen this? Can we have a test for this?

Reply:
I haven't personally seen it in the wild, but I based the above on [this](https://github.com/github/codeql/blob/main/python/ql/src/semmle/python/objects/ObjectAPI.qll#L204) similar construction in `ObjectAPI.qll`, so I _assume_ there's a reason for it. (I added the check that the file contains Python statements, because we don't want random non-Python files to affect the status of the Python files in the same directory.)

That being said, I'm not sure we can test this without modifying the extractor. I tried renaming `main.py` to have a non-`.py` extension, and I can include it directly using `--filter:include=...`, but even though the file appears in the database, it seems we're not interpreting it as Python code. My best guess is that it is being filtered out by a more restrictive "is this a script" check in the extractor, which seems unfortunate.

I'm also thinking that the above should read `this.getExtension().matches("py%")`, since there are many Python-related files that have extensions starting with `py`, and we certainly don't want to have those affecting the analysis, even if they _do_ contain Python statements.

Reply:
Ok, I think it is fine to have this structure in place.

Reply:
Oh, I do this every now and then if I write a small helper script in Python, that just needs to look like an executable. With the `#!/usr/bin/env python` everything works out anyway :+1:

OHH, but that is covered by the `exists(Comment c |` case... So I guess this is for `python my-script-without-extension`, which does seem strange.

> I'm also thinking that the above should read `this.getExtension().matches("py%")`, since there are many Python-related files that have extensions starting with `py`, and we certainly don't want to have those affecting the analysis, even if they _do_ contain Python statements.

Hmm, is this about `.pyc`, `.pyo` or `.pyd` files? I'm under the impression that since they contain byte-code, that wouldn't be valid Python statements.

(This last point is not super import to me, just curious)

Reply:
Indeed, not all `.py?` files contain Python code, so those would be filtered out anyway. I was thinking more of stub files like `.pyi` which _do_ contain valid Python, but not code that is actually executed as a script.

(I also found [this](http://dcjtech.info/topic/python-file-extensions/) random list of Python extensions. There are more than I had expected.)

Reply:
Good point about `.pyi` :+1: I personally didn't know about `.py3` usage either :smile: